src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs

```diff
@@ -33,53 +33,46 @@ public sealed class CryptClientCreator
         /// <summary>
         /// Create a CryptClient instance.
         /// </summary>
-        /// <param name="bypassQueryAnalysis">The bypass query analysis flag.</param>
-        /// <param name="encryptedFieldsMap">The encrypted fields map.</param>
-        /// <param name="kmsProviders">The kms providers.</param>
-        /// <param name="schemaMap">The schema map.</param>
+        /// <param name="cryptClientSettings">Crypt client settings.</param>
         /// <returns>The CryptClient instance.</returns>
-        public static CryptClient CreateCryptClient(
-            bool? bypassQueryAnalysis,
-            IReadOnlyDictionary<string, BsonDocument> encryptedFieldsMap,
-            IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> kmsProviders,
-            IReadOnlyDictionary<string, BsonDocument> schemaMap)
+        public static CryptClient CreateCryptClient(CryptClientSettings cryptClientSettings)
         {
-            var helper = new CryptClientCreator(bypassQueryAnalysis, encryptedFieldsMap, kmsProviders, schemaMap);
+            var helper = new CryptClientCreator(cryptClientSettings);
             var cryptOptions = helper.CreateCryptOptions();
             return helper.CreateCryptClient(cryptOptions);
         }
 #pragma warning restore
         #endregion
 
-        private readonly bool? _bypassQueryAnalysis;
-        private readonly IReadOnlyDictionary<string, BsonDocument> _encryptedFieldsMap;
-        private readonly IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> _kmsProviders;
-        private readonly IReadOnlyDictionary<string, BsonDocument> _schemaMap;
+        private readonly CryptClientSettings _cryptClientSettings;
 
-        private CryptClientCreator(
-            bool? bypassQueryAnalysis,
-            IReadOnlyDictionary<string, BsonDocument> encryptedFieldsMap,
-            IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> kmsProviders,
-            IReadOnlyDictionary<string, BsonDocument> schemaMap)
+        private CryptClientCreator(CryptClientSettings cryptClientSettings)
         {
-            _bypassQueryAnalysis = bypassQueryAnalysis;
-            _encryptedFieldsMap = encryptedFieldsMap;
-            _kmsProviders = Ensure.IsNotNull(kmsProviders, nameof(kmsProviders));
-            _schemaMap = schemaMap;
+            _cryptClientSettings = cryptClientSettings ?? new CryptClientSettings();
         }
 
         private CryptClient CreateCryptClient(CryptOptions options)
         {
-            return CryptClientFactory.Create(options);
+            var result = CryptClientFactory.Create(options);
```

Comment:
Additionally I think we need to modify the logic inside CryptClientFactory with creating an empty CryptClient first and then add a global try catch with disposing created client if any failed. I will think about implementation details in the scope of my current changes

Reply:
this logic should be different since it's almost exactly the same logic as already presents in the bindings in `handle.Check`: https://github.com/mongodb/libmongocrypt/blob/master/bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs#L49. The only difference that for this case we don't have `code` and `status` from c level. It happens because for this particular case it's a bit harder to determine whether we need to trigger exception or no on the c level, but the idea is the same. Exception should be: `new CryptException(Library.StatusType.MONGOCRYPT_STATUS_ERROR_CLIENT, -1, "..")`

Reply:
Also, notice that we don't have to call Dispose in this case because `ReleaseHandle` that we implement in SafeHandle is guaranteed to be called by GC https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.releasehandle?view=net-6.0#remarks

Reply:
>I am not sure about that, validating against ExtraOptions settings seems like a more highlevel validation than CryptClientFactory.

Effectively we do here configuring shared library. ExtraOptions just provides conditions for this. The factory in this case is responsible for either getting valid cryptClient or throw exception. We should not be dependent on any c level related behavior or expectations from that level

>As CryptClientFactory only validates that all native setup methods where set, before mongocrypt_init is called

csfle downloading also happens before `mongocrypt_init`. The only reason why c level doesn't throw himself is because we don't pass `IsCsfleRequired` value.

Reply:
>We should not wait for GC if we don't need the client anymore.

if this error happens, then all driver instance has been failed, so I'm not sure how important to clear resources immediately. Given that a handle is a local variable, it will be a first target for GC and it's guaranteed to be refreshed

>That makes me think that CryptClientFactory.Create needs try/catch block.

this is what I meant here https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/173/files#r875114967. This should be a part of bindings though. I'm ok with current implementation and relying on ReleaseHandle, but I don't mind if you add try catch there too.

Reply:
I have added `try catch` since probably it's better to do it explicitly and in time